### PR TITLE
Add enum support

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ?

```php
enum PostStatus: string
{
    case DRAFT = 'draft';
    case PUBLISHED = 'published';
}

class Post
{

    public function __construct(
        #[\Yiisoft\Validator\Rule\In(PostStatus::class)]
        private string $status
    )
    {
    }
}
```

### CODE REVIEW

**Comment by samdark** [comment]
Also need to cover new code with tests.

**Comment by samdark** [comment]
Likely need an ability to verify against names and non-backed enums.

**Comment by samdark** [comment]
`#[InEnum(PostStatus::class)]`. Could also set `names: true` to force using names for backed enums. For non-backend ones, names will be used.

File: tests/Rule/InEnumTest.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+use Yiisoft\Validator\Rule\InEnumHandler;
+use Yiisoft\Validator\Tests\Rule\Base\DifferentRuleInHandlerTestTrait;
+use Yiisoft\Validator\Tests\Rule\Base\RuleTestCase;
+use Yiisoft\Validator\Tests\Rule\Base\RuleWithOptionsTestTrait;
+use Yiisoft\Validator\Tests\Rule\Base\SkipOnErrorTestTrait;
+use Yiisoft\Validator\Tests\Rule\Base\WhenTestTrait;
+use Yiisoft\Validator\Tests\Support\Data\Enum\BackedEnumStatus;
+use Yiisoft\Validator\Tests\Support\Data\Enum\EnumStatus;
+
+final class InEnumTest extends RuleTestCase
```
Comment (xepozz) [inline tests/Rule/InEnumTest.php:18]: Could you also add tests for `int` backed enums?

**Review by xepozz — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ## 2.0.0 under development
 
+- Enh #734: Add enum support (@samdark)
```
Comment (vjik) [inline CHANGELOG.md:5]:
```suggestion
- New #734: Add `InEnum` rule (@samdark)
```

File: src/Rule/InEnum.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        return 'inEnum';
+    }
+
+    /**
+     * Get a set of values to check against.
+     *
+     * @return iterable A set of values.
+     */
+    public function getValues(): iterable
```
Comment (vjik) [inline src/Rule/InEnum.php:103]:
```suggestion
    public function getValues(): array
```

File: src/Rule/InEnumHandler.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+final class InEnumHandler implements RuleHandlerInterface
+{
+    public function validate(mixed $value, object $rule, ValidationContext $context): Result
+    {
+        if (!$rule instanceof InEnum) {
+            throw new UnexpectedRuleException(InEnum::class, $rule);
+        }
+
+        $result = new Result();
+        if ($rule->isNot() === ArrayHelper::isIn($value, $rule->getValues(), $rule->isStrict())) {
```
Comment (vjik) [inline src/Rule/InEnumHandler.php:27]: `in_array()` is enoguh, not need `ArrayHelper`.

File: src/Rule/InEnum.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+/**
+ * Defines validation options to check that the value is one of the values (or names) contained in an enum of the
+ * specified class.
+ * If the {@see In::$not} is set, the validation logic is inverted and the rule will ensure that the value
+ * is NOT one of them.
+ *
+ * In case of the validated value being a list, the order of values is important.
+ *
+ * Nested arrays are supported in the validated value (the order of values in
+ * lists must match, the order of keys in associative arrays is not important).
```
Comment (vjik) [inline src/Rule/InEnum.php:29]: hm, seems, array values is not supported...

File: src/Rule/InEnum.php [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    use SkipOnEmptyTrait;
+    use SkipOnErrorTrait;
+    use WhenTrait;
+
+    /**
+     * @param string $class Class of the enum to user.
+     * @param bool $useNames Whether to use names for backend enums instead of value.
+     * @param bool $strict Whether the comparison to each value in the set is strict:
+     *
+     * - Strict mode uses `===` operator meaning the type and the value must both match.
```
Comment (vjik) [inline src/Rule/InEnum.php:48]: Handler don't use `===` or `==` operators. `in_array()` is used.

File: src/Rule/InEnum.php [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+     * - {@link https://www.php.net/manual/en/language.types.type-juggling.php}
+     *
+     * Defaults to `false` meaning non-strict mode is used.
+     * @param bool $not Whether to invert the validation logic. Defaults to `false`. If set to `true`, the value must NOT
+     * be among the list of {@see $values}.
+     * @param string $message Error message when the value is not in a set of value.
+     *
+     * You may use the following placeholders in the message:
+     *
+     * - `{attribute}`: the name of the attribute.
```
Comment (vjik) [inline src/Rule/InEnum.php:63]:
```suggestion
     * - `{property}`: the name of the attribute.
```

Comment (vjik) [inline src/Rule/InEnum.php:63]: Add `{Property}` also

File: src/Rule/InEnum.php [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+     * See {@see SkipOnErrorInterface}.
+     * @param Closure|null $when A callable to define a condition for applying the rule.
+     * See {@see WhenInterface}.
+     *
+     * @psalm-param SkipOnEmptyValue $skipOnEmpty
+     * @psalm-param WhenType $when
+     */
+    public function __construct(
+        private string $class,
+        private bool $useNames = false,
```
Comment (vjik) [inline src/Rule/InEnum.php:76]: Should we throw exception when non-typed enums used with disabled `useNames` option?

**Review by vjik — CHANGES_REQUESTED** [review]

File: tests/Rule/InEnumTest.php (hunk 1 above)
Comment (samdark) [reply tests/Rule/InEnumTest.php:19]: Done.

File: CHANGELOG.md (hunk 2 above)
Comment (samdark) [reply CHANGELOG.md]: Done.

File: src/Rule/InEnum.php (hunk 8 above)
Comment (samdark) [reply src/Rule/InEnum.php:76]: No, I think it's OK to ignore this case.

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 ## 2.0.0 under development
 
+- New #734: Add enum support (@samdark)
```
Comment (vjik) [inline CHANGELOG.md:5]:
```suggestion
- New #734: Add `InEnum` rule (@samdark)
```

File: src/Rule/InEnum.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * Defines validation options to check that the value is one of the values (or names) contained in an enum of the
+ * specified class.
+ * If the {@see In::$not} is set, the validation logic is inverted and the rule will ensure that the value
+ * is NOT one of them.
+ *
+ * In case of the validated value being a list, the order of values is important.
```
Comment (vjik) [inline src/Rule/InEnum.php:26]:
```suggestion
```

File: src/Rule/InEnumHandler.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        if (!$rule instanceof InEnum) {
+            throw new UnexpectedRuleException(InEnum::class, $rule);
+        }
+
+        $result = new Result();
+        if ($rule->isNot() === in_array($value, $rule->getValues(), $rule->isStrict())) {
+            $result->addError(
+                $rule->getMessage(),
+                [
```
Comment (vjik) [inline src/Rule/InEnumHandler.php:29]: Fix mutant here.

**Review by vjik — CHANGES_REQUESTED** [review]

**Review by vjik — APPROVED** [review]

**Pull request merged**